### ISSUE TITLE
Strange behaviors with Greedy and defaults

### ISSUE THREAD

**Issue by MusicOnline** [Rapptz/discord.py#1791]
## Description

If `Greedy` is used as the final parameter, it can return an empty list.
However, when `Greedy` is used before the final parameter, it can fail.

With a command defined like this:
```py
@bot.command()
async def foo(ctx, a: commands.Greedy[int], *b):
    await ctx.send(f"{a} | {b}")
```

Invoking with `?foo 5` works, with output: `[5] | ()`

Invoking with `?foo A` breaks, with error: `Converting to "int" failed for parameter "a".`
Expected output would be `[] | ('A')`.

## With `typing.Optional`

Combining `Greedy` with `typing.Optional` is unable to circumvent this.
- When `Optional[Greedy[int]]`, the error raised when defining the command is:
```
TypeErrpr: Optional[t] requires a single type. Got <discord.ext.commands.converter._Greedy object at 0x0EACA050>.
```
- When `Greedy[Optional[int]]`, the error raised when defining the command is:
```
TypeError: Greedy[...] expects a type.
```
(this issue does not propose using `typing.Optional` as the solution!)

## Use case
Semi-practical:
```py
@commands.guild_only()
@bot.command()
async def broadcast(ctx, channels: commands.Greedy[discord.TextChannel], *, message):
    if not channels:
        channels = ctx.guild.text_channels
    for ch in channels:
        await ch.send(message)
```

**Comment by MusicOnline** [comment]
> `Greedy` parameters can also be made optional by specifying an optional value.

This was mentioned [here in the documentation](https://discordpy.readthedocs.io/en/rewrite/ext/commands/commands.html#greedy).

However, if the `Greedy` parameter is the final parameter (with no convertable args passed), an empty list is always returned, despite setting a default.
![capture1](https://user-images.githubusercontent.com/39178127/50265529-fe358980-045a-11e9-935f-31676db8a2cd.PNG)

If the `Greedy` parameter is not the final parameter and with an immutable default (with no convertable args passed), two things can happen:
![capture2](https://user-images.githubusercontent.com/39178127/50265593-42288e80-045b-11e9-9787-0ea99ee9cf68.PNG)

(command signatures displayed are of the latest revision of the rewrite branch, for the issue regarding inaccurate command signatures, see #1653)

**Comment by MusicOnline** [comment]
Resolved as of commit dc8aa7c35b5e6399d6ed2dcde50eb59ab8abc744.
